Added custom array example with documentation.

marvin-hansen opened the pull request:
Added example code with a comprehensive Readme that covers the bulk of the custom array concepts and how they relate to each other.

In response to discussion: https://github.com/diesel-rs/diesel/discussions/4157

However, the Readme needs review and the tests need a fix for a serialization error.

marvin-hansen commented:
Quick update:

* Tests are fixed
* Readme got edited and checked
* Added the clippy lint check to the GH action so it runs together with all the others on CI

Only thing I could not figure out was how to add the tests to CI.

Should be straight forward to review this PR.

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+
+To serialize and deserialize a custom Enum, you write a custom toSql and fromSql implementation. Luckily, this is
+straightforward.
+
```
I personally would like to add a paragraph here that explains what these impl do.

`ToSql` describes how to serialize a given rust type (`Self`) as sql side type (first generic argument) for a specific database backend (second generic argument). It needs to translate the type into the relevant wire protocol in that case. For postgres/mysql enums that's just the enum value as ascii string, but in general it's depended on the database + type.

`FromSql` describes how to deserialize a given rust type (`Self`) as sql side type (first generic argument) for a specific database backend (second generic argument). It need to translate from the relevant wire protocol to the rust type. For postgres/mysql enums that just means matching on the as ascii string, but in general it's depended on the database + type.

marvin-hansen replied:
That is a great explanation. I added it as it is to the Custom enum paragraph.

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,892 @@
+}
+```
+
+It is worth mentioning is that the sql_type refers to the wrapper struct generated by Diesel and stored in the schema.rs
````
> wrapper struct

We call that type sql side type in our documentation. (At least we try to be consistent)

marvin-hansen replied:
Good to know

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,892 @@
+}
+```
+
+I cannot stress enough that it is paramount that the tuple type signature must match exactly the Postgres type signature defined in your up.sql. Ideally, you want to use the split view function in your IDE to have the up.sql 
````
Note for diesel-team/myself: We might want to look into generating that as part of the `schema.rs` file as well. Maybe just as `impl RecordLike { type Fields = (…)}` for record types and maybe for enums `impl EnumLike { const FIELDS: &str = &[…] }` 

(That would likely also result in diesel-cli generating the relevant SQL type for the enum here as well)

To be clear: That's nothing that you need to address as this PR, that's just me thinking loud.

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,892 @@
+```
+
+I cannot stress enough that it is paramount that the tuple type signature must match exactly the Postgres type signature defined in your up.sql. Ideally, you want to use the split view function in your IDE to have the up.sql 
+in one pane and the the toSql implementation in another pane, both side by side, to double check 
````
```suggestion
in one pane and the the ToSql implementation in another pane, both side by side, to double check 
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+that the number and types match. 
+If the type or number of types mismatch, you will get a compiler error telling you that somehow either 
+the number of fields don’t match or that the type of the fields don’t match. 
+Also, because the write_tuple expects values, you have to call either to_owned() or clone() on any referenced data.
```
I think it should work with `&self.version`, etc as well, but I've not checked that (as of writing this comment)

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+2) CreateService: For create operations
+3) UpdateService: For update operations
+
+The Service and CreateService have identical fields. Also, both types require a primary key annotation in addition to
+the table name. The UpdateService, however, has each field wrapped into an option type. When the option type is Some,
+Diesel knows that this field needs updating. If the field is set to None, Diesel ignores it.
```
It might be worth to link the "All about inserts/updates" guides here for the relevant parts?

marvin-hansen replied:
Done.

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,892 @@
+    }
+```
+
+Run the compile and see that it works. Now, if you were to remove the type annotation <i64> from get_result, you get a
+compile error saying that trait FromSql is not implemented.
````
Just to add a bit more context here: That's sometimes requires because there can be more than one `FromSql` impl for the same rust type and there can be more than one `FromSql` type for the same sql type. That gives a lot flexibility how you actually map data between your database and your rust code, but it requires that you be explicit about these types.

weiznich reviewed (commented):
Thanks for working on this. This is already a really great starting point. 

I left a bunch of comments on various things. Some of them are just more information that might be helpful for you and others that read that example. Some are minor stylistic comments and a few are things that we should likely improve. If something is unclear just let me know and I will try to explain it with more details.

That written: If that example is merged I will try to convert it into a guide on the web page. It's already a really good starting point, it mostly needs styling/linking for that. This hopefully makes it much more discoverable how to perform that kind of extensions.

weiznich commented:
> Only thing I could not figure out was how to add the tests to CI.

The tests are automatically picked up: https://github.com/diesel-rs/diesel/actions/runs/10411801607/job/28836874448?pr=4169#step:27:356. I'm not sure why they fail but I can have a look at them next week if you want.

marvin-hansen commented:
@weiznich Thank you so much for the substantial commentary. T

he bulk of your comments are relatively easy to fix. I still trying to figure out the details about how Diesel handles custom schema, so your comments here are invaluable and I most likely will update the Readme  once I updated the code so that everything resides in the custom schema. Still learning... 

I have fixed the failing tests. Apparently, the CI picks up the test automatically and I simply wasn't running a migration prior to test execution. Once I added an embedded migration, the test seems to pass for now. 
On that topic, I also added a section about Integration testing to the Readme to complete the document in terms of covering all the important topics. 

Moving the content to the Website would be great! 

I look into all the details over the weekend and update the PR

marvin-hansen commented on examples/postgres/custom_arrays/src/model/service/service_impl.rs:
```diff
@@ -0,0 +1,147 @@
+        db: &mut Connection,
+        param_service_id: i32,
+    ) -> QueryResult<bool> {
+        match service
+            .filter(service_id.eq(param_service_id))
+            .select(online)
+            .first::<bool>(db)
+        {
+            Ok(res) => Ok(res),
+            Err(e) => Err(e),
+        }
```
Correct.  This is a left over from returning another error. Good catch

marvin-hansen commented on examples/postgres/custom_arrays/src/model/protocol_type/mod.rs:
```diff
@@ -0,0 +1,47 @@
+            b"GRPC" => Ok(ProtocolType::GRPC),
+            b"HTTP" => Ok(ProtocolType::HTTP),
+            b"UDP" => Ok(ProtocolType::UDP),
+            _ => Ok(ProtocolType::UnknownProtocol),
```
ok, makes sense. I added a DatabaseError to convey that the DB contains something that isnt expected.

```rust
fn from_sql(bytes: PgValue<'_>) -> deserialize::Result<Self> {
        match bytes.as_bytes() {
            b"UnknownProtocol" => Ok(ProtocolType::UnknownProtocol),
            b"GRPC" => Ok(ProtocolType::GRPC),
            b"HTTP" => Ok(ProtocolType::HTTP),
            b"UDP" => Ok(ProtocolType::UDP),
            _ => Err(DatabaseError(
                DatabaseErrorKind::SerializationFailure,
                Box::new(format!(
                    "Unrecognized enum variant: {:?}",
                    String::from_utf8_lossy(bytes.as_bytes())
                )),
            )
            .into()),
        }
    }
```

marvin-hansen commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+searches in the public schema for a table. If that fails, Postgres returns an error that the table does not exists (in the default schema).
+Because of the search path mechanism, you only have to declare the schema name and consistently prefix thetables in the schema with the schema name and that’s it.
+
+It is important to know that a custom schema only applies to tables. Custom types and enum still
+reside in the default public schema. 
```
Ok. I worked on this but got a bunch or errors so i removed parts.
It seems to be working now after having correctly annotated the wrapper types with the schema.
Updating the Readmen to reflect that.

marvin-hansen commented:
Update:

* Changed dotenv to dotenvy; (see [comment](https://github.com/diesel-rs/diesel/pull/4169#discussion_r1719584842)) 
* Moved Endpoint impl into type definition file. (See [comment](https://github.com/diesel-rs/diesel/pull/4169#discussion_r1719596303))
* Switched example code and tests from PGConnectionPool to single PGConnection. (See [comment](https://github.com/diesel-rs/diesel/pull/4169#discussion_r1719587567))
* Moved all custom types into custom schema smdb. See [comment](https://github.com/diesel-rs/diesel/pull/4169#discussion_r1719613646)
* Added a DatabaseError to PgProtocolType in from_sql implementation to convey that the DB contains something that isn't expected. See [comment](https://github.com/diesel-rs/diesel/pull/4169#discussion_r1719595463). 

Updating the Readme next.

marvin-hansen commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+2) You want to group a bunch of fields i.e. to store a configuration file with nested fields
+3) You specifically don’t want to model relations
+
+In the case of a service endpoint, it really is as simple as every service has one or more endpoints, but these are
```
Done.

marvin-hansen commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+* The array type follows the usual convention type[].
+* The NOT NULL attribute means that the array itself must be set; Values inside the array still might be null.
+
+The Diesel team decided that array values are nullable by default because there so many things that may go wrong when
+dealing with array values thus making them nullable by default prevents a few potential crashes. Conversely, it also
```
Correct. Fixing it.

marvin-hansen commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,892 @@
+pub mod smdb {
+    pub mod sql_types {
+        #[derive(diesel::query_builder::QueryId, Clone, diesel::sql_types::SqlType)]
+        #[diesel(postgres_type(name = "service_endpoint", schema = "smdb"))]
```
True. As mentioned before, I've removed stuff so this is a left-ofter.
That said, this commented really helped me to understand that I have to annotated all wrapper types with the custom schema and then stuff worked.

marvin-hansen commented:
Okay, the Readme and tests are all updated with all suggestions being resolved.

The remaining two conflicts are:

.github/workflows/ci.yml
diesel/src/pg/query_builder/copy/copy_to.rs

The first one is probably that I have added the example to the ci config.
The second refers to a suppressed linter I have added to make clippy on CI pass.

Please check if either one is okay, and if not, just change whatever needs to be changed. 

Lastly, I really don't know why so many CI jobs got stuck in limbo. I've read on HN that GH had a major outage recently, but really no idea what's happening there.

Either way, please double check the Readme and if there are no major objections, feel free to merge anytime.

weiznich commented:
Thanks for the update :+1: I've rebased the PR to resove the merge conflicts. I'm now going over the changes to perform another review.

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+### External primary key: INTEGER type
+
+In case you need to know the specific value of the primary key and you need to know it before inserting the data, you
+have to assign unique primary keys before inserting data and you have to use the INTEGE type (or any of the many integer
```
```suggestion
have to assign unique primary keys before inserting data and you have to use the INTEGER type (or any of the many integer
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+Lastly, you also want to add a type alias for a postgres database connection in the lib file because that gives you the
+freedom to swap between a normal single connection or a connection pool without updating your type implementation or
+tests.
+The connection pool type alias has ben uncommented, but the type signature of the pooled connection makes it obvious why
```
```suggestion
The connection pool type alias has been uncommented, but the type signature of the pooled connection makes it obvious why
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+
+## Rust Types
+
+In total, you implement three Rust types:
```
```suggestion
In total, you need implement three Rust types:
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+
+Notice, the Postgres types are stored in Postgres therefore you are not seeing them in the generated schema. Only tables
+will show up in the generates schema. Furthermore, you will need a wrapper struct to map from Rust to the Postgres type.
+For the ServiceEndpoint, Diesel already generated a matching wrapper struct with the correct annotations. The service
+table then uses that wrapper struct “ServiceEndpoint” instead of the native Postgres type to reference the custom type
+in the endpoints array.
```
```suggestion
For the ServiceEndpoint, Diesel already generated a matching zero sized SQL type struct with the correct annotations. The service
table then uses that SQL type `ServiceEndpoint` instead of the native Postgres type to reference the custom type
in the endpoints array.
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+
+### Rust Enum: ProtocolType
+
+Diesel needs a wrapper struct to store a custom Enum in Postgres, so let’s add one:
```
```suggestion
Diesel needs a zero sized SQL type struct to represent a custom Enum in Postgres, so let’s add one:
```

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,1162 @@
+pub struct PgProtocolType;
+```
+
+It is important that you add both, sql_type and postgres_type, otherwise insert will fail.
+Furthermore, it is important to add the database schema ("smdb") otherwise Postgres fails to find the type and aborts an
+operation on that type.
````
```suggestion
It is important to add the database schema ("smdb") and type name ("protocol_type") otherwise Postgres fails to find the type and aborts an
operation on that type.
```

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,1162 @@
+
+```rust
+#[derive(SqlType)]
+#[diesel(sql_type = protocol_type)]
````
```suggestion
```

as that's not needed

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+Furthermore, it is important to add the database schema ("smdb") otherwise Postgres fails to find the type and aborts an
+operation on that type.
+
+Next, let’s define the actual Enum. Bear in mind to use the wrapper struct as sql_type on the Enum:
```
```suggestion
Next, let’s define the actual Enum. Bear in mind to use the SQL type struct as sql_type on the Enum:
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+It’s worth pointing out that you are dealing with three types in total:
+
+1) protocol_type: The Postgres Enum type
+2) PGProtocolType: The Wrapper struct that refers to the Postgres Enum type
```
```suggestion
2) PgProtocolType: The SQL type struct that refers to the Postgres Enum type
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+
+1) protocol_type: The Postgres Enum type
+2) PGProtocolType: The Wrapper struct that refers to the Postgres Enum type
+3) ProtocolType: The Rust Enum that refers to the wrapper struct PGProtocolType
```
```suggestion
3) ProtocolType: The Rust Enum that maps to the SQL type struct PgProtocolType
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+a convoluted way to say that the database type mismatches the Rust type. When you encounter a consulted trait error
+message, make sure to check:
+
+1) Do I have a wrapper struct?
+2) Does my a wrapper struct derives SqlType?
+3) Does my wrapper type has both, sql_type and postgres_type declared?
+4) Are sql_type and postgres_type both referring to the correct Postgres type?
+5) Does my Rust type refers to the wrapper struct type?
```
```suggestion
1) Do I have a SQL type struct?
2) Does my a SQL type struct derive SqlType?
3) Does my SQL type has a `#[diesel(postgres_type(_)]` attribute declared?
4) Is the  `#[diesel(postgres_type(_)]` attribute referring to the correct Postgres type?
5) Does my Rust type refers to the SQL type struct?
```

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,1162 @@
+
+```rust 
+#[derive(Debug, Clone, FromSqlRow, AsExpression, PartialEq, Eq)]
+#[diesel(sql_type=crate::schema::smdb::sql_types::ServiceEndpoint)]
````
```suggestion
#[diesel(sql_type = crate::schema::smdb::sql_types::ServiceEndpoint)]
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+Postgres represent a custom type as an anonymous typed tuple. Therefore, you have to map a struct to a typed tuple and
+back.
+
+Let’s start with the toSql implementation to store the Rust Endpoint struct as typed tuple.
```
```suggestion
Let’s start with the ToSql implementation to store the Rust Endpoint struct as typed tuple.
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+the number of fields don’t match or that the type of the fields don’t match.
+Also, because the write_tuple expects values, you have to call either to_owned() or clone() on any referenced data.
+
+The from_sql reverses the process by converting a Postgres typed tuple back into a Rust struct.
```
```suggestion
The FromSql reverses the process by converting a Postgres typed tuple back into a Rust struct.
```

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,1162 @@
+```
+
+Similar to the serialization process, it is paramount that the type annotation
+match exactly the one used in toSql and the type definition in up.sql.
````
```suggestion
match exactly the one used in ToSql and the type definition in up.sql.
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+
+It’s worth repeating that you are dealing with three types in total:
+
+1) service_endpoint: The Postgres custom type
+2) ServiceEndpoint: The wrapper struct generated by Diesel that refers to the Postgres type service_endpoint. Also, only
+   the wrapper struct carries the postgres_type annotation in addition to the an sql_type annotation.
+3) Endpoint: The Rust struct with an sql_type annotation referring to the wrapper struct ServiceEndpoint
```
```suggestion
1) service_endpoint: The Postgres custom type
2) ServiceEndpoint: The SQL type struct generated by Diesel that refers to the Postgres type service_endpoint. Also, only
   the SQL type struct carries the postgres_type annotation.
3) Endpoint: The Rust struct with an sql_type annotation referring to the SQL type struct ServiceEndpoint
```

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,1162 @@
+
+```rust
+#[derive(Debug, Clone, Queryable, Selectable)]
+#[diesel(table_name= crate::schema::smdb::service,  primary_key(service_id))]
````
```suggestion
#[diesel(table_name = crate::schema::smdb::service, primary_key(service_id))]
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+}
+
+#[derive(Debug, Clone, Queryable, Insertable)]
+#[diesel(table_name= crate::schema::smdb::service,  primary_key(service_id))]
```
```suggestion
#[diesel(table_name = crate::schema::smdb::service, primary_key(service_id))]
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+}
+
+#[derive(Debug, Clone, Queryable, Insertable, AsChangeset)]
+#[diesel(table_name= crate::schema::smdb::service)]
```
```suggestion
#[diesel(table_name = crate::schema::smdb::service)]
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+        param_service_id: i32,
+        param_online: bool,
+    ) -> QueryResult<()> {
+        match diesel::update(service.filter(service_id.eq(param_service_id)))
+            .set(online.eq(param_online))
+            .returning(Service::as_returning())
+            .get_result(db)
+        {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
```
```suggestion
        diesel::update(service.filter(service_id.eq(param_service_id)))
            .set(online.eq(param_online))
            .execute(db)?;
        Ok(())
```

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+        db: &mut Connection,
+        param_service_id: i32,
+    ) -> QueryResult<bool> {
+        match service.find(param_service_id).first::<Service>(db) {
+            Ok(_) => Ok(true),
+            Err(_) => Ok(false),
+        }
```
```suggestion
        Ok(service.find(param_service_id).first::<Service>(db).optional()?.is_some())
```

weiznich commented on examples/postgres/custom_arrays/README.md:
````diff
@@ -0,0 +1,1162 @@
+To check if a service is online, we again lean on the generated DSL:
+
+```rust
+    pub fn check_if_service_id_exists(
+        db: &mut Connection,
+        param_service_id: i32,
+    ) -> QueryResult<bool> {
+        match service.find(param_service_id).first::<Service>(db) {
+            Ok(_) => Ok(true),
+            Err(_) => Ok(false),
+        }
+    }
````
That seems to be copy + pasted from above. We should likely use the correct function from the example here

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+        services: &[i32],
+    ) -> QueryResult<(bool, Option<String>)> {
+        for id in services {
+            match Service::check_if_service_id_online(db, *id) {
+                Ok(res) => {
+                    if !res {
+                        return Ok((false, Some(format!("Service {} is offline", id))));
+                    }
+                }
+                Err(e) => return Err(e),
+            };
```
```suggestion
            let res = Service::check_if_service_id_online(db, *id)?;
            if !res {
                return Ok((false, Some(format!("Service {} is offline", id))));
            }
```

marvin-hansen replied:
That is from check_all_services_online; I've replaced the sample in the Readme with the actual implementation.
Good catch.

weiznich commented on examples/postgres/custom_arrays/README.md:
```diff
@@ -0,0 +1,1162 @@
+
+Database integration tests become flaky when executed in parallel usually because of conflicting read / write
+operations.
+While modern database systems can handle concurrent data access, test tools with assertions not so much.
+That means, test assertions start to fail seemingly randomly when executed concurrently.
+There are only very few viable options to deal with this reality:
+
+* Don't use parallel test execution
+* Only parallelize test per isolated access
+* Do synchronization and test the actual database state before each test and run tests as atomic transactions
```
There is a simple trick here that enables you to write concurrent database tests: Use a dangling transaction. Diesel uses that extensively for it's own test-suite. What that means is that you perform the following steps:

* Open a new connection per test
* Directly start a new transaction (Diesel provides the `Connection::begin_test_transaction` function for this)
* Perform any test code (including schema setup for postgres)
* Rollback the transaction

With that you end up with mostly isolated views on the database, where each test does not see the changes from other tests. (Well you can observe different id's etc, but that's all)

marvin-hansen replied:
Okay, I actually added another testsuite to the example that implements the parallel testing via begin_test_transaction. That way, people have the freedom to adopt either the test group parallel test style or the per test transaction testing style.

Very good point, I've updated the Readme.

weiznich reviewed (commented):
I've went over everything again and it looks much better now. Most comments are only minor grammar/spelling/consistency suggestions. I'm happy to do the cleanup on this next week if you want.

marvin-hansen commented:
Thanks a lot for the consistency checks as I didn't had the time to do it. Overall, I already committed the bulk of corrections.

 I want to take a closer look at your suggestion to encapsulate tests into a transaction. I get the idea of basically piggy backing on the two phase commit mechanism inherent to transactions. I just have to write and run some sample tests to see how this works out in practice.

I think if that works out, it's sensible to update the test suite and the Readme. 

Background is, I'm building and testing my Rust mono repo with Bazel and my CI runs already hundreds of tests in parallel. However,  for DB integration tests at scale, the aborted transaction trick may work out and speeds things up further.

marvin-hansen commented:
Thanks a lot for the consistency checks as I didn't had the time to do it. Overall, I already committed the bulk of corrections.

 I want to take a closer look at your suggestion to encapsulate tests into a transaction. I get the idea of basically piggy backing on the two phase commit mechanism inherent to transactions. I just have to write and run some sample tests to see how this works out in practice.

I think if that works out, it's sensible to update the test suite and the Readme. 

Background is, I'm building and testing my Rust mono repo with Bazel and my CI runs already hundreds of tests in parallel. However,  for DB integration tests at scale, the aborted transaction trick may work out and speeds things up further.

marvin-hansen commented:
@weiznich 

Update summary:
* Readme was updated with all those minor corrections you have flagged
* Updated run_db_migration in lib by removing unnecessary check 
* Added revert_db_migration to lib; it's used to cleanup the DB after sequential testing. 
* Added parallel test suite using test transaction, as you laid out in a previous comment. 
* Updated Readme section on testing.  

Just one question: Does the test transaction aborts automatically when it goes out of scope?
I did not find any abort or rollback method, but somehow all tests run flawlessly and I wonder why? 

Either way, please review these changes and if there 
is nothing in the way, feel free to merge whenever ready.

weiznich commented:
Thanks for the update. I will try to do a final review (and likely some additional minor adjustments based on that) this week. Otherwise I think this is fine now.

> Just one question: Does the test transaction aborts automatically when it goes out of scope?
I did not find any abort or rollback method, but somehow all tests run flawlessly and I wonder why?

It relays on the fact that you drop the connection itself at the end of the test. This will also drop any uncommited changes for that connection.

marvin-hansen commented:
@weiznich 

Thank you for the detailed corrections and the links to the docs in the Readme. Also, thank you for getting the parallel tests in shape. While things worked locally on my box, apparently on CI each test really needs an unique connection. 

It was my first try with the  dangling transaction test style so there is stil a bit of learning ahead. Anyways, just want to thank you for your time and effort to get this PR done.

weiznich commented:
>  While things worked locally on my box, apparently on CI each test really needs an unique connection.

The problem there was that you likely had already applied the migrations manually. The tests run against an empty database and postgres do not allow to modify the schema from more than one connection at once. That's worked around by having the schema modification itself in the test transaction as well. So it doesn't get commited and there are no problem with concurrent access.

Thanks again for writing this up. I think that's now a good starting point for at least two guides: One for custom types and one for testing. Now I only need to find the capacity to do the last bit of polishing to put it on the web page.

weiznich commented:
> As for the guide, I don't have the capacity to do that as my project progresses, but it might be an option to open an issue with a call for contribution with some guidance and some links to an example. This would be a great first PR for new contributors.

I'm already really grateful that you wrote this PR. I will find the time at some point for the necessary work to turn this into a full guide, I just wanted to manage expectations that this likely won't be the case tomorrow.
